src/SourceCode.Clay.Buffers/IO/BufferSessionStream.cs

```diff
@@ -0,0 +1,283 @@
+#region License
+
+// Copyright (c) K2 Workflow (SourceCode Technology Holdings Inc.). All rights reserved.
+// Licensed under the MIT License. See LICENSE file in the project root for full license information.
+
+#endregion
+
+using SourceCode.Clay.Buffers;
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace SourceCode.Clay.IO
+{
+    internal sealed class BufferSessionStream : Stream
+    {
+        #region Fields
+
+        private readonly object _lock;
+        private BufferSession _session;
```

Comment:
Can this not be `readonly`

Reply:
Nope, if it is the inner `_isDisposed` field is not updated during `Dispose()` and this allows the buffer to be returned to the pool multiple times.